### PR TITLE
Update lastnpe EEA to 2.4.0

### PR DESCRIPTION
**Pull request by lsiepel**
Resolves: https://github.com/openhab/openhab-addons/issues/10396

_The reason is that the new version fixes that Optional.orElse can return null values (based on the parameter).
There is quite some code using Optional.orElse expecting the result to be non-null._

In most cases i wrapped the optional in `Objects.requireNonNull(..)` and in some cases i removed the optional.

At time of writing bindings A-D have been adjusted and i like to get some feedback before proceeding to fix all. 
Edit: Moved on to bindings A - K

JFTR: to be merged after 4.2 release as i cannot test _all_ bindings.

### CODE REVIEW

**Comment by J-N-K** [comment]
The problem is that we can't do proper null annotations with EEA. Unfortunately there is no will on Eclipse side to fix that. We would need something like "non-null if argument is non-null". This also applies to `Map.compute` and similar methods.

**Comment by lsiepel** [comment]
> The problem is that we can't do proper null annotations with EEA. Unfortunately there is no will on Eclipse side to fix that. We would need something like "non-null if argument is non-null". This also applies to `Map.compute` and similar methods.

Is there some kind of link to that discussion? And more important do you have some thoughts on how to proceed in relation to this PR?

**Comment by J-N-K** [comment]
https://github.com/lastnpe/eclipse-null-eea-augments/issues/116 and linked issues

https://bugs.eclipse.org/bugs/show_bug.cgi?id=575822

File: bundles/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDeviceRunnable.java
Comment (wborn) [inline bundles/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDeviceRunnable.java:92]: This looks nicer and its shorter for such constructs:

```suggestion
                    Objects.requireNonNullElse(e.getMessage(), "");
```

Comment (lsiepel) [reply bundles/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDeviceRunnable.java:92]: Thanks, fixed (alsop on other places)

**Comment by morph166955** [comment]
I can't validate this any time soon for the AndroidTV/PhillipsTV bit.  Is this a Java 17 update or would this still be backwards compatible to Java 11?  I'm about to kill off the 3.x compatibility but I haven't yet so I'm trying to keep that in mind still.

**Comment by lsiepel** [comment]
> I can't validate this any time soon for the AndroidTV/PhillipsTV bit. Is this a Java 17 update or would this still be backwards compatible to Java 11? I'm about to kill off the 3.x compatibility but I haven't yet so I'm trying to keep that in mind still.

As far as i know this should be java 11 compatible. Maybe 4.2 is a good point to move on anyway.

**Review by holgerfriedrich — APPROVED** [review]
LGTM (for KNX)

File: bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/devices/bridge/LongPolling.java
Comment (david-pace) [inline bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/devices/bridge/LongPolling.java:171]: `contentAsString` will ultimately end up in `Gson::fromJSON` which should handle `null` gracefully (see https://github.com/google/gson/blob/99cc4cb11f73a6d672aa6381013d651b7921e00f/gson/src/main/java/com/google/gson/Gson.java#L1167).

I would therefore prefer to add `@Nullable` annotations to the `content` parameters of `onLongPollComplete()` and `handleLongPollResponse()` instead of changing the logic here.

File: bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/discovery/ThingDiscoveryService.java
Comment (david-pace) [inline bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/discovery/ThingDiscoveryService.java:217]: I suppose this is necessary because the library cannot properly detect automatically that this is non-null? I would prefer to wait until the library does this properly instead of making the code more complicated than it has to be.

**Review by david-pace — CHANGES_REQUESTED** [review]
Thanks, I have two alternative proposals and a question.

**Comment by andrewfg** [comment]
For the Hue binding the change is not really necessary since the argument is a constant (and therefore definitely not null). But the change won't do any harm either.

**Review by mherwege — CHANGES_REQUESTED** [review]
Ohterwise LGTM for mybmw and nikohomecontrol.

**Review by ecdye — APPROVED** [review]
LGTM for Omnilink

File: bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/discovery/ThingDiscoveryService.java
Comment (lsiepel) [reply bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/discovery/ThingDiscoveryService.java:217]: Unfortunately the library is not expected to be updated with that enhancement anytime soon. Agree that there is room for improvement.
This PR aligns the dependency with the other repo’s and also brings some fixes

File: bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/devices/bridge/LongPolling.java
Comment (lsiepel) [reply bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/devices/bridge/LongPolling.java:171]: Agree, much better.

File: bundles/org.openhab.binding.mybmw/src/test/java/org/openhab/binding/mybmw/internal/dto/charge/ChargingStatisticsTest.java
Comment (wborn) [reply bundles/org.openhab.binding.mybmw/src/test/java/org/openhab/binding/mybmw/internal/dto/charge/ChargingStatisticsTest.java:123]: Due to limitations the null analysis has no idea what happens in method implementations being called. It only considers annotations on other methods and null checks in the current method. It would be very time consuming for the analysis to check each every method calling another method dozens of levels deep including JVM, framework and library code. 🙃

**Review by theater — APPROVED** [review]
Checked the changes in Solax binding. They're not breakable for sure as the logic is made to always return non-null. LGTM (if you need me to review the other changes just let me know but it will take more time) :)

**Review by florian-h05 — APPROVED** [review]
LGTM for JS Scripting.

**Review by mlobstein — APPROVED** [review]
LGTM for radiothermostat

File: bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/devices/bridge/LongPolling.java
Comment (david-pace) [inline bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/devices/bridge/LongPolling.java:169]: Please revert to the original version without the curly braces since we now have the Sonar issue `remove useless curly braces around statement` (rule `java:S1602`)

**Review by david-pace — CHANGES_REQUESTED** [review]
Found one remaining minor issue that was introduced through unnecessary curly braces.

File: bundles/org.openhab.automation.jrubyscripting/src/main/java/org/openhab/automation/jrubyscripting/internal/JRubyScriptEngineConfiguration.java
Comment (wborn) [inline bundles/org.openhab.automation.jrubyscripting/src/main/java/org/openhab/automation/jrubyscripting/internal/JRubyScriptEngineConfiguration.java:382]: It's the default value and assigning it again wastes CPU cycles.

```suggestion
        private @Nullable String value;
```

File: bundles/org.openhab.binding.androidtv/src/main/java/org/openhab/binding/androidtv/internal/protocol/philipstv/PhilipsTVConnectionManager.java
Comment (wborn) [inline bundles/org.openhab.binding.androidtv/src/main/java/org/openhab/binding/androidtv/internal/protocol/philipstv/PhilipsTVConnectionManager.java:234]: I think you still need to run `spotless:apply`:

```suggestion
            this.username = Objects.requireNonNullElse(configMap.get("username"), "");
```

File: bundles/org.openhab.binding.mielecloud/src/main/java/org/openhab/binding/mielecloud/internal/webservice/api/json/Type.java
Comment (wborn) [inline bundles/org.openhab.binding.mielecloud/src/main/java/org/openhab/binding/mielecloud/internal/webservice/api/json/Type.java:46]: This can be a one-liner using `Objects.requireNonNullElse(..)`

File: bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/api/data/ModuleType.java
Comment (wborn) [inline bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/api/data/ModuleType.java:222]: This can be a one-liner using `Objects.requireNonNullElse(..)`

File: bundles/org.openhab.binding.sagercaster/src/main/java/org/openhab/binding/sagercaster/internal/caster/SagerWeatherCaster.java
Comment (wborn) [inline bundles/org.openhab.binding.sagercaster/src/main/java/org/openhab/binding/sagercaster/internal/caster/SagerWeatherCaster.java:93]:
```suggestion
    private @Nullable SagerPrediction prevision;
```

File: bundles/org.openhab.binding.sagercaster/src/main/java/org/openhab/binding/sagercaster/internal/caster/SagerWeatherCaster.java
Comment (wborn) [inline bundles/org.openhab.binding.sagercaster/src/main/java/org/openhab/binding/sagercaster/internal/caster/SagerWeatherCaster.java:348]: You can probably reduce the lines of code using `Objects.requireNonNullElse` in this file too.

File: bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/dto/SurePetcareDevice.java
Comment (wborn) [inline bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/dto/SurePetcareDevice.java:47]: Maybe it's easy to annotate the class and its super class with `@NonNullByDefault` so `@NonNull` can be removed?

**Review by wborn — COMMENTED** [review]
Thanks for taking the time to do this! :+1: You're almost there. :slightly_smiling_face:
I added a few more minor comments:

**Review by druciak — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/devices/bridge/LongPolling.java
Comment (lsiepel) [reply bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/devices/bridge/LongPolling.java:169]: Don't see that issue/waning in vscode, or where should i see this?
Removed the braces anyway, thanks.

File: bundles/org.openhab.automation.jrubyscripting/src/main/java/org/openhab/automation/jrubyscripting/internal/JRubyScriptEngineConfiguration.java
Comment (lsiepel) [reply bundles/org.openhab.automation.jrubyscripting/src/main/java/org/openhab/automation/jrubyscripting/internal/JRubyScriptEngineConfiguration.java:382]: Fixed

File: bundles/org.openhab.binding.androidtv/src/main/java/org/openhab/binding/androidtv/internal/protocol/philipstv/PhilipsTVConnectionManager.java
Comment (lsiepel) [reply bundles/org.openhab.binding.androidtv/src/main/java/org/openhab/binding/androidtv/internal/protocol/philipstv/PhilipsTVConnectionManager.java:234]: Fixed

File: bundles/org.openhab.binding.mielecloud/src/main/java/org/openhab/binding/mielecloud/internal/webservice/api/json/Type.java
Comment (lsiepel) [reply bundles/org.openhab.binding.mielecloud/src/main/java/org/openhab/binding/mielecloud/internal/webservice/api/json/Type.java:46]: doh... fixed.

File: bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/api/data/ModuleType.java
Comment (lsiepel) [reply bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/api/data/ModuleType.java:222]: fixed

File: bundles/org.openhab.binding.sagercaster/src/main/java/org/openhab/binding/sagercaster/internal/caster/SagerWeatherCaster.java
Comment (lsiepel) [reply bundles/org.openhab.binding.sagercaster/src/main/java/org/openhab/binding/sagercaster/internal/caster/SagerWeatherCaster.java:93]: Fixed

File: bundles/org.openhab.binding.sagercaster/src/main/java/org/openhab/binding/sagercaster/internal/caster/SagerWeatherCaster.java
Comment (lsiepel) [reply bundles/org.openhab.binding.sagercaster/src/main/java/org/openhab/binding/sagercaster/internal/caster/SagerWeatherCaster.java:348]: Unfortunately not. Calling a method on a nullable is generating a compiler warning. Or do you mean something differnet then:
`return Objects.requireNonNullElse(this.prevision.getForecast(), UNDEF);`

File: bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/dto/SurePetcareDevice.java
Comment (lsiepel) [reply bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/dto/SurePetcareDevice.java:47]: I prefer to leave it as is. Changing this to `@NonNullByDefault` would also mean i have to look after all other members without knowing anything from this binding. Maybe something to revisit in the future.

File: bundles/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/internal/handler/SatelSystemHandler.java
Comment (lsiepel) [reply bundles/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/internal/handler/SatelSystemHandler.java:85]: It won't compile otherwise. See OP.

File: bundles/org.openhab.binding.sagercaster/src/main/java/org/openhab/binding/sagercaster/internal/caster/SagerWeatherCaster.java
Comment (wborn) [reply bundles/org.openhab.binding.sagercaster/src/main/java/org/openhab/binding/sagercaster/internal/caster/SagerWeatherCaster.java]: Yes you're right! Didn't look closely. :detective:

File: bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/devices/bridge/LongPolling.java
Comment (david-pace) [reply bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/devices/bridge/LongPolling.java]: Try installing the `SonarLint` extension, it is a brilliant tool that helped me to find and fix many issues, and as a "side effect" to even learn a lot about Java and other languages :+1:

**Review by david-pace — APPROVED** [review]
Thanks, looks good now for `boschshc` :+1:

File: bundles/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/internal/handler/SatelSystemHandler.java
Comment (druciak) [reply bundles/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/internal/handler/SatelSystemHandler.java:85]: This is IMO step back. We sacrifice readability in the name of using broken tool.

**Comment by jimtng** [comment]
> _The reason is that the new version fixes that Optional.orElse can return null values (based on the parameter). There is quite some code using Optional.orElse expecting the result to be non-null._

Please help me understand the exact problem that this PR is trying to solve. Is it the fact that `Optional.orElse(<this can return null>)` is allowed and can return `null`?

**Comment by wborn** [comment]
In a nutshell, this PR updates the code so it can compile with the latest EEA library that had some annotation updates. This includes:

> 1. Is it the fact that Optional.orElse(<this can return null>) is allowed and can return null?

File: bundles/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/internal/handler/SatelSystemHandler.java
Comment (lsiepel) [reply bundles/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/internal/handler/SatelSystemHandler.java:85]: Some improvements come at a cost. I understand your pov. I mark this as resolved as this is not blocking this PR

Comment (wborn) [reply bundles/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/internal/handler/SatelSystemHandler.java:85]: Yes there are some slight inconveniences but the big advantage still remains that the tool helps to prevent many NPEs.

**Review by druciak — APPROVED** [review]

**Review by weymann — APPROVED** [review]
Fine for sensorcommunity

**Review by wborn — APPROVED** [review]
Thanks a lot! :+1:

**Review by bern77 — APPROVED** [review]
LGTM for helioseasycontrol

**Comment by lsiepel** [comment]
@openhab/add-ons-maintainers now 4.3 development is started, would be nice to have this merged sooner then later.

**Comment by wborn** [comment]
Looks like some new code got added that no longer compiles with the newer EEAs.
Can you update that too @lsiepel?  After that we can merge the PR.

**Comment by lsiepel** [comment]
@openhab/add-ons-maintainers ping

**Pull request merged**
